### PR TITLE
[RPA-1116] Variables autocomplete functionality

### PR DESCRIPTION
**Pull request by Aeri4a**
## Description

image

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

---

I agree to the terms of the RunBotics Contributor License Agreement.

### CODE REVIEW

File: runbotics/runbotics-orchestrator-ui/src/src-app/views/process/ProcessBuildView/Modeler/ActionFormPanel/widgets/AutocompleteWidget/AutocompleteWidget.tsx [hunk 1]
```diff
@@ -38,38 +82,88 @@ const AutocompleteWidget: FC<AutocompleteWidgetProps> = ({
...
+                onClick={handleExternalOpen}
+            >
+                <If
+                    condition={isExternalOpen}
+                    else={<KeyboardArrowUpIcon/>}
+                >
+                    <KeyboardArrowDownIcon/>
+                </If>
+            </StyledButton>
+        </StyledBox>
```
Comment (preludium) [inline runbotics/runbotics-orchestrator-ui/src/src-app/views/process/ProcessBuildView/Modeler/ActionFormPanel/widgets/AutocompleteWidget/AutocompleteWidget.tsx]: adjust input adornment instead
https://mui.com/material-ui/react-text-field/#input-adornments

File: runbotics/runbotics-orchestrator-ui/src/src-app/views/process/ProcessBuildView/Modeler/ActionFormPanel/widgets/AutocompleteWidget/AutocompleteWidget.tsx [hunk 2]
```diff
@@ -1,9 +1,20 @@
...
+import KeyboardArrowDownIcon from '@mui/icons-material/KeyboardArrowDown';
+import KeyboardArrowUpIcon from '@mui/icons-material/KeyboardArrowUp';
 import { TextField, Autocomplete } from '@mui/material';
 
+import If from '#src-app/components/utils/If';
+
+import { StyledBox, StyledButton } from './AutocompleteWidget.styles';
 import { AutocompleteWidgetProps } from './AutocompleteWidget.types';
 
+const inputDetailsDefault = {
```
Comment (preludium) [inline runbotics/runbotics-orchestrator-ui/src/src-app/views/process/ProcessBuildView/Modeler/ActionFormPanel/widgets/AutocompleteWidget/AutocompleteWidget.tsx]: upper case and low dash if constant

File: runbotics/runbotics-orchestrator-ui/src/src-app/views/process/ProcessBuildView/Modeler/ActionFormPanel/widgets/AutocompleteWidget/AutocompleteWidget.tsx [hunk 3]
```diff
@@ -15,18 +26,51 @@ const AutocompleteWidget: FC<AutocompleteWidgetProps> = ({
...
+    const [inputDetails, setInputDetails] = useState(inputDetailsDefault);
+    const [isExternalOpen, setIsExternalOpen] = useState(false);
     const optionValues = React.useMemo(
         () => ({
             'ui:options': Object.values(autocompleteOptions).map((option) => option.value),
         }),
         [autocompleteOptions]
     );
 
+    const analyzeInput = (input: string) => {
```
Comment (preludium) [inline runbotics/runbotics-orchestrator-ui/src/src-app/views/process/ProcessBuildView/Modeler/ActionFormPanel/widgets/AutocompleteWidget/AutocompleteWidget.tsx]: more specific function name and created variables

**Review by preludium — CHANGES_REQUESTED** [review]
Having issues handling clear on select close? Try to use close related props https://mui.com/material-ui/api/autocomplete/#Autocomplete-prop-clearOnBlur

File: runbotics/runbotics-orchestrator-ui/src/src-app/views/process/ProcessBuildView/Modeler/ActionFormPanel/widgets/AutocompleteWidget/AutocompleteWidget.tsx [hunk 4]
```diff
@@ -15,18 +25,51 @@ const AutocompleteWidget: FC<AutocompleteWidgetProps> = ({
...
             'ui:options': Object.values(autocompleteOptions).map((option) => option.value),
         }),
         [autocompleteOptions]
     );
 
+    const checkInputMatches = (input: string) => {
+        const variableRegex = /[#$]\{([^{}]*)\}?/g;
+        if (!input) return INPUT_AUTOCOMPLETE_DEFAULT;
+
+        const matches = [...input.matchAll(variableRegex)];
```
Comment (preludium) [inline runbotics/runbotics-orchestrator-ui/src/src-app/views/process/ProcessBuildView/Modeler/ActionFormPanel/widgets/AutocompleteWidget/AutocompleteWidget.tsx]: more like `variables` imo. I assume that the right side of this line is to retrieve all variables in autocomplete string

File: runbotics/runbotics-orchestrator-ui/src/src-app/views/process/ProcessBuildView/Modeler/ActionFormPanel/widgets/AutocompleteWidget/AutocompleteWidget.tsx [hunk 5]
```diff
@@ -15,18 +25,51 @@ const AutocompleteWidget: FC<AutocompleteWidgetProps> = ({
...
+        if (!matches.length) return INPUT_AUTOCOMPLETE_DEFAULT;
+
+        const autocompleteResult = matches.find(match => !(optionValues['ui:options'].includes(match[0])));
```
Comment (preludium) [inline runbotics/runbotics-orchestrator-ui/src/src-app/views/process/ProcessBuildView/Modeler/ActionFormPanel/widgets/AutocompleteWidget/AutocompleteWidget.tsx]: what is a structure of `matches` right now? 2D list?

Comment (preludium) [reply runbotics/runbotics-orchestrator-ui/src/src-app/views/process/ProcessBuildView/Modeler/ActionFormPanel/widgets/AutocompleteWidget/AutocompleteWidget.tsx]: so here you look for a single variable that is not completed at the moment? if so then then apply this to you variable name

File: runbotics/runbotics-orchestrator-ui/src/src-app/views/process/ProcessBuildView/Modeler/ActionFormPanel/widgets/AutocompleteWidget/AutocompleteWidget.tsx [hunk 6]
```diff
@@ -38,25 +81,62 @@ const AutocompleteWidget: FC<AutocompleteWidgetProps> = ({
         setOpen(false);
     };
 
-    const handleChange = (event: any, newValue: any) => {
+    const handleChange = (event: any, newValue: string) => {
+        setInputAutocompleteState(checkInputMatches(newValue));
         onChange(newValue || undefined);
     };
+
+    const handleAutocomplete = (event: any, newValue: string) => {
```
Comment (preludium) [inline runbotics/runbotics-orchestrator-ui/src/src-app/views/process/ProcessBuildView/Modeler/ActionFormPanel/widgets/AutocompleteWidget/AutocompleteWidget.tsx]: tbh for sure there is type for such event

File: runbotics/runbotics-orchestrator-ui/src/src-app/views/process/ProcessBuildView/Modeler/ActionFormPanel/widgets/AutocompleteWidget/AutocompleteWidget.tsx [hunk 7]
```diff
@@ -38,25 +81,62 @@ const AutocompleteWidget: FC<AutocompleteWidgetProps> = ({
...
+
     return (
         <Autocomplete
             fullWidth
-            autoSelect
-            autoComplete
             disableCloseOnSelect={false}
-            autoHighlight
-            open={open}
+            disableClearable={true}
```
Comment (preludium) [inline runbotics/runbotics-orchestrator-ui/src/src-app/views/process/ProcessBuildView/Modeler/ActionFormPanel/widgets/AutocompleteWidget/AutocompleteWidget.tsx]: `={true}` is redundant for boolean props

**Review by preludium — CHANGES_REQUESTED** [review]

**Review by preludium — APPROVED** [review]

**Comment by Aeri4a** [comment]
Gateway & Autocomplete still to fix after rebase [now working after fixes]

**Pull request merged**